### ISSUE TITLE
Feature request: allow to restrict app password data access

### ISSUE THREAD

**Issue by j-ed** [nextcloud/server#2866]
## Expected behavior
If a user creates individual app passwords, it should be possible to restrict to which data the app should have access to and what kind of access should be allowed.

Example:
```
App         RO  RW  CREATE  DELETE  UPLOAD
----------  --  --  ------  ------  ------
Calendar    X
Contacts        X
Files                                 X          
```

## Current behavior
If a user creates individual app passwords, he couldn't restrict to which data the app should have access. 

### Steps to reproduce
1. Login as as a user
2. Choose Personal -> App passwords

#### Server Configuration

OS: Linux 3.2.82
Web server: Apache2 2.4.25
Database: MariaDB 5.5.53
PHP version: 5.6.29
Nextcloud version: 10.0.2

#### Client Configuration

Browser: Mozilla Firefox 50.1.0
Operating system: Windows 7

**Comment by MariusBluem** [comment]
See: https://github.com/nextcloud/server/pull/719

What is the current plan/state? @icewind1991

**Comment by nickvergessen** [comment]
The problem is that this is not really that easy.

E.g. you wouldn't want your activity reader to do anything right?
But it needs access to all of those to display the related events, get the contact names for displaying and file access for handling of paths in activities.

So while it is nice in theory, we need to make this easily checkable and somehow a way to correctly deal with this for other apps.

**Comment by benyanke** [comment]
@nickvergessen It would seem that to implement this, a permissions framework would need to be built into the core of NC, at which point apps could register the permissions they have available, which would be namespaced with the app name, etc, etc.


A framework I'm using on my own projects is https://github.com/spatie/laravel-permission. While it's deeply integrated with laravel, so not very suitable for this, perhaps a similar model could be used: create a list of permissions - as you mentioned, separate out read vs write permissions in each section, and then assign to users.

**Comment by nickvergessen** [comment]
Yeah something similar to the bruteforce protection we added in 12 (or was it 11?).
https://github.com/nextcloud/server/blob/0bccd5a0d9cda570f405e4f6add4d0a27ec3aee4/core/Controller/LoginController.php#L219

**Comment by szaimen** [comment]
I suppose this issue is still valid? If not, please close this issue!

**Comment by j-ed** [comment]
Yes, you're right. It is still not possible to restrict the data access.

**Comment by riedel** [comment]
Actually it makes sense to go even further. IMHO it would make sense to use the sharing functionality to share a file or subdirectory with an "app password". Particularly if one needs to upload or download data for  a longer timeframe in a multiuser environment keeping any password on disk or in memory with such wide permissions does not much lessen the attack surface. 

Context: I am currently looking on a good way how to instruct users to include nextcloud files in projects in their HPC project

**Comment by InRiPa** [comment]
I'm using Android Apps, which access a specific App from nextcloud (e.g. app for calendar, app for tasks, ..). It would be nice to restrict the access of given "app password account" to the exact purpose.

**Comment by sunjam** [comment]
As an aside, also worth considering implementing scoped access for oauth, rather than always granting full access to Nextcloud itself when using it for authentication with external services.
- https://github.com/nextcloud/server/issues/26233

**Comment by K0-RR** [comment]
Linking related posts on the forum:
https://help.nextcloud.com/t/app-specific-passwords-folder-file-permissons/2162
https://help.nextcloud.com/t/limiting-access-of-external-apps-clients-to-a-limited-folder-or-even-file/18939

**Comment by jsalatiel** [comment]
It is a pity such a great feature is not given proper priority.

**Comment by sbrudenell** [comment]
> The problem is that this is not really that easy.
> 
> E.g. you wouldn't want your activity reader to do anything right? But it needs access to all of those to display the related events, get the contact names for displaying and file access for handling of paths in activities.

Not every case needs such complexity. 

For example I want to sync my keepass db, and only that one file, to my phone.

It would be totally reasonable to start with file/folder based permissions for this feature. Let users manage the complexity for now. The permissions framework can evolve later into a fancier, feature-based framework.

**Comment by joshtrichards** [comment]
Related:

- #9117 
- #35262
- #17339

**Comment by sunjam** [comment]
@joshtrichards wondering if this would be done in combination scopes with Oauth (as is done in Slack).  See
- https://github.com/nextcloud/server/issues/26233

**Comment by DFINITYManu** [comment]
What I would like is an app password just for contacts and calendar, so the device I sync my PIM info with doesn't have access to files at all.

Would be similarly good to have an app password just for News, so the device I do my news reading doesn't have access to my PIM or file data at all.

**Comment by NiklasBeierl** [comment]
I'd also be a huge fan of this feature. As mentioned by other folks, you can grant limited file access by creating separate users for separate clients / devices. But for calendars / contacts, this would be a pain in the rear. 

One place where I have seen this implemented quite nicely is [sogo](https://github.com/Alinto/sogo).
When creating an app password you get a dialogue like this:
![Image](https://github.com/user-attachments/assets/300fdeed-4607-4c0f-9350-b31790368032)

(Further distinguishing between calendar and contacts would be a plus)

**Comment by saywebsolutions** [comment]
Yes this feature is desperately needed. Consider wanting to use an app's API, for instance the cookbook API. It would be ideal if you could limit that credential to only be able to access the Cookbook API instead of risking the whole nextcloud account being pwned.

**Comment by riedel** [comment]
> One place where I have seen this implemented quite nicely is [sogo](https://github.com/Alinto/sogo).

@NiklasBeierl : can you point me to the place where this SoGo feature is implemented? (docs, source code). It seems super great (background: we are using sogo webmail with an underlying cyrus with ldap auth and it would be a charm if we didn't have to pass the original clear text password of the user to SoGo.)

Having said that. Nextcloud should have much less difficulties, since they already have app passwords/tokens implemented and this is only about scoping them. I guess here the user dialog is the easy part, but rather passing different sub-user-id through the whole stack could be a hell.

**Comment by NiklasBeierl** [comment]
@riedel I looked into it and it turns out this is not a feature of sogo but of [mailcow dockerized](https://github.com/mailcow/mailcow-dockerized). Couldn't find dedicated documentation, but if you search the repo for "app passwords" you get to some php code.

**Comment by My1** [comment]
at least in terms of files one HUGE step forward would already be the idea to scope an app password to a folder which basically becomes the new "root folder" for that application, so you can make a folder for e.g. your password manager, or automatic backups from a website and it can do all the things it normally can do except that all starts within that folder.

**Comment by crispyduck00** [comment]
@My1 That would be really great.
Of course it would not cover all use cases, but it would already make it possible to have one app password per folder. If multiple folders in different locations are needed, it would still be possible to create multiple app passwords.

I strongly prefer the principle of least privilege, meaning each app/password should only have access to what is actually required.

I initially thought I could work around this for my backup/sync apps by using public shares and accessing them via /public.php/dav, but I found that using Basic Auth with different apps triggers the bruteforce protection very quickly.

My current workaround is using separate user accounts. This is not really ideal, but still better than giving an app—which should, for example, only back up to a specific folder—write access to everything the main user can access.

Implementing to at least set a specific root folder per app password would be a significant improvement.

**Comment by riedel** [comment]
Just to add one comment regarding work arounds: those seem fine for home installations but for large corporate installations (which might even have higher security demands), arbitrary users are not possible. Also weird patterns easily break things or trigger security mechanisms (which result in complicated discussions with the admins about what one is trying to do)

**Comment by AceDenghar** [comment]
Please,

In this time of large data leaks it's a non-sense that this is not yet implemented.

Apps password is a good first step but should not be unfinished as it is.

There are many apps that can connect to Nextcloud and it's a good thing for Nextcloud's popularity and usages.

It's also a good reflex for users to use apps passwords.

But we can't be sure of what these applications can do and even if using other login/password that the principal one, they can do almost everything with our datas.

Adding these options could be a very good point to secure them.

Sorry for my english and thank you for the fantastic job you've done on Nextcloud ❤️

Ace

**Comment by makua104** [comment]
I ran into this today trying to limit the WebDAV access of Keepassium on iOS for access to my password database in Nextcloud. My goal was to avoid giving the app full read/write permission to every folder my Nextcloud user has.

As mentioned by other commenters it appears that using password protected share links is an OK workaround. I found an old reddit thread discussing how to do it [here](https://www.reddit.com/r/NextCloud/comments/q2j85a/how_to_access_a_shared_link_password_via_webdav/) and it's also kind of documented in the Nextcloud docs [here](https://docs.nextcloud.com/server/latest/user_manual/en/files/access_webdav.html#accessing-public-shares-over-webdav) and [here](https://docs.nextcloud.com/server/20/user_manual/en/files/access_webdav.html#accessing-public-shares-over-webdav)

Summary of the solution for e.g. a scoped folder:
1. Create the folder
2. Create a public password protected share for the folder with e.g. `https://cloud.example.com/s/yxcFKRWBJqYYzp4` as the share link
3. Use e.g. `https://cloud.example.com/public.php/webdav` as the WebDAV connection URL (could probably change in future versions, the two doc links appear to use different URL paths) enter the part of the previous url after `/s/` so in our case `yxcFKRWBJqYYzp4` as the username and then the share password as the password for the user.
4. WebDAV client only has access to the inside of the folder

**Comment by archont00** [comment]
Thanks a lot - let me re-write since I was struggling a bit with understanding of the steps:

1. NC: Create a folder.
2. NC: Create an external share of that folder with e-mail and password.
3. NC: Display the QR code for the share and scan it with phone camera to find out the direct URL with the token. E.g. `https://cloud.example.com/s/yxcFKRWBJqYYzp4`.
4. In WebDAV client app (Keepass2Android in my case), set
    - URL: `https://cloud.example.com/public.php/webdav`
    - Username: `yxcFKRWBJqYYzp4`
    - Password: the share password
5. The WebDAV client should display the shared folder content

**Comment by AceDenghar** [comment]
Thank you for this workaround ! It will be usefull 🥰

But should be centralized in apps password when used by apps...

We should clearly distinguish apps and users access.

But it's good to know !

**Comment by pmilosev** [comment]
I assume the issue is non-standardised permissions across Nextcloud apps.
But for start it would be nice to limit which Nextcloud apps the app password grants access to.
This is already solved in the Guests app:

<img width="1131" height="275" alt="Image" src="https://github.com/user-attachments/assets/ad7cf01a-d5a1-4f5b-8b93-384c65bf5e0a" />

**Comment by sunjam** [comment]
Sorry to be broken record, three years later...

> We should clearly distinguish apps and users access.

That would be handled through scoped oauth.  Currently Nextcloud only grants full access, with zero granularity, to all issued oauth tokens.  See [the actual Nextcloud documentation](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/oauth2.html), which states:

> Nextcloud OAuth2 implementation currently does not support scoped access. This means that every token has full access to the complete account including read and write permission...

- https://github.com/nextcloud/server/issues/26233



OAuth Scopes is a mechanism in OAuth 2.0 to limit an application's
access to a user's account. An application can request one or more scopes,
this information is then presented to the user in the consent screen, and
the access token issued to the application will be limited to the scopes
granted.

Full details at https://oauth.net/2/scope/

> Example:
> 
> ```
> App         RO  RW  CREATE  DELETE  UPLOAD
> ----------  --  --  ------  ------  ------
> Calendar    X
> Contacts        X
> Files                                 X          
> ```

This exact example from the original post here is what Oauth scoped access is for.  Please do not take my word for it, and do some research.  Hope this helps!  You could define _all of the following within one single oauth token:_

Calendar: read-only
Contacts: read/write
Files: upload-only

And even add more things like:
list
download
edit metadata
share
delete
manage folders
So instead of one blanket “full access” token, Nextcloud could issue tokens with narrow capabilities for each app and each action, as defined for each specific token.
